### ISSUE TITLE
Support `devEngines` field

### ISSUE THREAD

**Issue by GeoffreyBooth** [pnpm/pnpm#8153]
### Contribution

- [ ] I'd be willing to implement this feature ([contributing guide](https://github.com/pnpm/pnpm/blob/main/CONTRIBUTING.md))

### Describe the user story

I help maintain Node.js, including Corepack. We’ve been searching for a way to provide interoperability with npm for Corepack users. npm has said that they won’t support Corepack or its `packageManager` field, but they’re open to the idea of a _similar_ field, that has some different characteristics that npm considers vital: semver ranges, flexibility on what to do when validation fails, validation for both the package manager and the runtime. This has led to a standardization effort with the OpenJS Foundation to define a new field called `devEngines`: https://github.com/openjs-foundation/package-metadata-interoperability-collab-space/issues/15. [npm is in the process of implementing support for it](https://github.com/openjs-foundation/package-metadata-interoperability-collab-space/issues/15#issuecomment-2140751719) (other than the `download` sub-field).

Once npm adds support, the plan is to propose support for this field to Corepack, as it would contain all the same information as `packageManager`. Corepack would probably continue to support `packageManager` for backward compatibility, but the hope is that if npm, Corepack and other package managers all standardize on support for `devEngines`, then users could be assured that they can validate the package managers used for their projects.

### Describe the solution you'd like

pnpm should also implement `devEngines` per https://github.com/openjs-foundation/package-metadata-interoperability-collab-space/issues/15, or as much of it as makes sense for pnpm (for example, I assume they might follow npm and ignore the download parts).

### Describe the drawbacks of your solution

I understand if pnpm wants to wait and see npm go first, to see this _actually_ ship; and then pnpm can copy or share code from npm.

I can’t promise that Corepack will definitely support `devEngines`, but the field’s whole purpose is to provide a configuration for both npm and Corepack to share so I expect it to happen.

### Describe alternatives you've considered

Continue to support `packageManager` and its downsides, as shown by https://github.com/pnpm/pnpm/issues/8087.

**Comment by zkochan** [comment]
I am OK to add support after npm adds support for it. We already support the `engines.pnpm` field which supports a range, unlike the `packageManager` field, so I don't think it is urgent.

> Continue to support packageManager and its downsides, as shown by https://github.com/pnpm/pnpm/issues/8087.

Regardless of this field, I don't think support of the `packageManager` field will go anywhere unless corepack will adopt `devEngines`.

**Comment by GeoffreyBooth** [comment]
> I am OK to add support after npm adds support for it. We already support the `engines.pnpm` field which supports a range, unlike the `packageManager` field, so I don’t think it is urgent.

Sounds good. The advantage of `devEngines.packageManager` is that it can control not just the version of the package manager, but _which_ package manager the user should be using. So if a new teammate checks out a project and runs `npm install`, npm would see `devEngines.packageManager.name === 'pnpm'` and error to the user telling them to use pnpm instead.

**Comment by zkochan** [comment]
@GeoffreyBooth I was working on adding packageManager support to pnpm: https://github.com/pnpm/pnpm/pull/8363

@jakebailey mentioned that the packageManager field also support a blob after a + sign to verify the correctness of the downloaded package manager's content.

However, we currently export two pnpm packages: pnpm and @pnpm/exe. @pnpm/exe is bundled with Node.js and pnpm is a plain js package. Hence, we don't have one integrity checksum, we have 2. 

I see that the devEngines.packageManager field allows an array. So if we were to support this new field, could we have two entries? One for the pnpm package with its integrity and another one for @pnpm/exe?

**Comment by fz6m** [comment]
npm has added support for `devEngines`, see [v10.9.0](https://github.com/npm/cli/releases/tag/v10.9.0)

**Comment by trivikr** [comment]
FYI: npm does not support `onFail=download` specification which is one of the user friendly features of [corepack](https://github.com/nodejs/corepack).

I've created reports on both npm as well as package metadata interoperability collab space:
* npm https://github.com/npm/cli/issues/8004
* package metadata interoperability collab space https://github.com/openjs-foundation/package-metadata-interoperability-collab-space/issues/33

**Comment by danielbayley** [comment]
> What is the status on this?

@kaaax0815 I’m working on it in #9707. Feel free to `checkout` and test, or just follow along :)

**Comment by ravindUwU** [comment]
Can this be closed? 🤔 To my understanding, PNPM supports Node.js/Deno/Bun runtime management via `devEngines.runtime` since [v10.14](https://github.com/pnpm/pnpm/releases/tag/v10.14.0).

**Comment by zkochan** [comment]
No, this cannot be closed because we still use the `packageManager` field instead of `devEngines.packageManager` for definining the pnpm version required by the project. The main problem for us is that we use npm CLI in some commands under the hood. For instance, we run `npm publish` from `pnpm publish`. If we'll add pnpm to devEngines.packageManager, then npm publish will throw an error.

I think these are our options:
1. if npm CLI would support some env variable to skip the devEngines check, pnpm could set that. I have started a discussion about it at npm: https://github.com/npm/rfcs/issues/830
2. pnpm could use npm's programmatic API instead. That's a lot of work.
3. we could put both npm and pnpm in the devEngines.packageManager field as it can be an array. This one might be the proper solution as both are used in reality. However, users would be able to run npm install instead of pnpm install.

**Comment by GeoffreyBooth** [comment]
Thank you @zkochan, this is very exciting to see!
